backend/src/main/kotlin/org/icpclive/service/TeamSpotlightService.kt

```diff
@@ -0,0 +1,131 @@
+            flowSettings.acceptedRunScore.takeIf(run.isAccepted) +
+            flowSettings.judgedRunScore.takeIf(run.isJudged) +
+            flowSettings.notJudgedRunScore.takeIf(!run.isJudged)
+}
+
+class TeamState(val teamId: Int, private val flowSettings: TeamSpotlightFlowSettings) : Comparable<TeamState> {
+    val score
+        get() = internalScore
+    private var internalScore = 0.0
+    private var causedRun: RunInfo? = null
+    val caused: KeyTeamCause
+        get() = causedRun?.let { RunCause(it.id) } ?: ScoreSumCause
+
+    fun addAccent(accent: TeamAccent) {
+        internalScore += accent.getScoreDelta(flowSettings)
+        if (accent is TeamRunAccent) {
+            causedRun = causedRun?.coerceAtMost(accent.run) ?: accent.run
+        }
+    }
+
+    override fun compareTo(other: TeamState): Int = when {
+        other.score > score -> 1
+        other.score < score -> -1
+        else -> (teamId - other.teamId)
+    }
+}
+
+class TeamSpotlightService(
+    val scope: CoroutineScope,
+    val settings: TeamSpotlightFlowSettings = TeamSpotlightFlowSettings()
+) {
+    private val mutex = Mutex()
+    private val queue = mutableSetOf<TeamState>()
+    private suspend fun getTeamInQueue(teamId: Int) = mutex.withLock {
+        queue.find { it.teamId == teamId } ?: TeamState(teamId, settings).also { queue += it }
+    }
+
+    fun getFlow(): Flow<KeyTeam> {
+        return flow {
+            while (true) {
+                val element = mutex.withLock {
+                    queue.minOrNull()?.also { queue.remove(it) }
+                }
+                if (element == null) {
+                    delay(1.seconds)
+                    continue
+                }
+                emit(KeyTeam(element.teamId, element.caused))
+            }
+        }
+    }
+
+    suspend fun run(
+        info: MutableStateFlow<ContestInfo>,
+        runs: Flow<RunInfo>,
+        scoreboard: Flow<Scoreboard>,
+        // analyticsMessage: Flow<AnalyticsMessage>
+    ) {
+        val runIds = mutableSetOf<Int>()
+        var contestInfo = info.value
+        var previousScoreboardPushTime = Duration.ZERO
+        merge(
+            tickerFlow(15.seconds).map { Trigger },
+            info,
+            runs
+        ).collect { update ->
+            when (update) {
+                is ContestInfo -> {
+                    contestInfo = update
+                }
+                is RunInfo -> {
+                    if (update.time + 60.seconds > contestInfo.currentContestTime) {
+                        if (update.isJudged || update.id !in runIds) {
+                            runIds += update.id
+                            getTeamInQueue(update.teamId).addAccent(TeamRunAccent(update))
+                        }
+                    }
+                }
+                is Trigger -> {
+                    if (contestInfo.currentContestTime - previousScoreboardPushTime > settings.scoreboardPushInterval) {
```

Comment:
Why not just setup trigger with correct interval? Do you think it should be possible to change it in runtime?